x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -0,0 +1,515 @@
+        }
+        
+        @Override
+        public Builder index(boolean index) {
+            if(index == false) {
+                throw new MapperParsingException("The field [" + name +
+                        "] cannot have index = false");                
+            }
+            return this;
+        }
+
+        public Builder numChars(int numChars) {
+            if ((numChars > MAX_NUM_CHARS_COUNT) || (numChars < 1)) {
+                throw new MapperParsingException("The number of characters for ngrams in field [" + name +
+                    "] should be in the range [1, " + MAX_NUM_CHARS_COUNT + "]");
+            }
+            this.numChars = numChars;
+            return this;
+        }
+        
+        public Builder ignoreAbove(int ignoreAbove) {
+            if (ignoreAbove < 0) {
+                throw new IllegalArgumentException("[ignore_above] must be positive, got " + ignoreAbove);
+            }
+            this.ignoreAbove = ignoreAbove;
+            return this;
+        }        
+          
+        
+        @Override
+        protected void setupFieldType(BuilderContext context) {
+            super.setupFieldType(context);
+            fieldType().setNumChars(numChars);
+            fieldType().setHasDocValues(true);
+            fieldType().setIndexOptions(IndexOptions.DOCS);                
+        }
+
+        @Override
+        public WildcardFieldType fieldType() {
+            return (WildcardFieldType) super.fieldType();
+        }
+
+        @Override
+        public WildcardFieldMapper build(BuilderContext context) {
+            setupFieldType(context);            
+            return new WildcardFieldMapper(
+                    name, fieldType, defaultFieldType, ignoreAbove, numChars,
+                    context.indexSettings(), multiFieldsBuilder.build(this, context), copyTo);
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+        @Override
+        public Mapper.Builder<?, ?> parse(String name, Map<String, Object> node, ParserContext parserContext)
+                throws MapperParsingException {
+            WildcardFieldMapper.Builder builder = new WildcardFieldMapper.Builder(name);
+            parseField(builder, name, node, parserContext);
+            
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String propName = entry.getKey();
+                Object propNode = entry.getValue();
+                if (propName.equals("num_chars")) {
+                    if (propNode == null) {
+                        throw new MapperParsingException("Property [numChars] cannot be null.");
+                    }
+                    builder.numChars(XContentMapValues.nodeIntegerValue(propNode));
+                    iterator.remove();
+                } else if (propName.equals("ignore_above")) {
```

Comment:
I'm unsure whether we need support for `ignore_above`. This field is actually well suited for long values?

Reply:
The field now has aggs support too, so a sensible safeguard is often to use ignore_above to tune out garbage from otherwise short fields. Admittedly that's a data quality issue perhaps best dealt with elsewhere but that's not always something you can rely on